### PR TITLE
SDL Input: Support more types of force feedback for controllers through SDL

### PR DESCRIPTION
**Pull request by adamdmoss**
Three new haptic modes tested with XBox 360 controller under Linux where no force-feedback was previously available in Dolphin.

### CODE REVIEW

**Comment by BhaaLseN** [comment]
Maybe @karolherbst could take a look? (see also PR #1801)

I think initializing ```m_effect``` in the ctor should be good enough, no need for a memset there.

**Comment by adamdmoss** [comment]
re: m_effect, I agree, but since the structure flips between a 0-type and e.g. a SINE type, and each type re-uses the same structure memory as a union, I thought explicit clears seem safer.  In case, say, a future version of SDL actually cares about any of the shared values being non-zero for a 0-type haptic (unlikely! but.)

**Comment by BhaaLseN** [comment]
You get a distinct ```m_effect``` instance for every effect, they are not shared among them. The only thing you have in the end is a populated effect with type set to 0 - and that shouldn't break anything (unless SDL is really doing odd things).
Infact, the other effects do that too (look at Constant and/or Ramp); they only set ```type```/```length``` when enabled, but set all other flags afterwards.

**Comment by karolherbst** [comment]
@BhaaLseN yeah, the m_effect commit from my PR fixes the issues I had. Also I think we should work on that inside one pr. The values seems interesting, but I won't have time to check them today. @adamdmoss you could base your commits on top of my branch: https://github.com/karolherbst/dolphin/tree/rumble and just add a commit with your values you think are good.

I think the "m_effect.leftright.small_magnitude = (Uint16)(state * 0xFFFF);" is troublesome, because it looks weird to use Uint16 here but Sint16 everywhere else.

**Comment by karolherbst** [comment]
I think we could use the LeftRight effect to create own good feeling effects, but trying to simulate or behave like the others might not be the right approach, because then we could simply use the other ones.

**Comment by BhaaLseN** [comment]
Not sure if ```small_magnitude``` is Uint16 ot Sint16, but it should be consistent either way.
Please do some (common) testing to see if you can find useful values that make the most of the three effects.

**Comment by karolherbst** [comment]
I think it is a good idea to have something really static there and control the effect strenght soley through the state and try to make the "periodical" effects as static as possible, so we won't disrubt any games (like imagine puzzle games and the player start to think the rumble effect has a hidden message or something)

**Comment by adamdmoss** [comment]
The LeftRight magnitudes are uint16, the rest are sint16.  Strange, I know, but the SDL test app for LeftRight also uses full 0xFFFF magnitude.

**Comment by adamdmoss** [comment]
I've tested the three effects as implemented and they're pretty similar, at least in magnitude, on my device.  I don't think it's particularly interesting for the 3 new effects to be used in tandem, it's more about wanting to include devices that just have 1 of the 3.

**Comment by adamdmoss** [comment]
And again, the memset is there to be explicit and surprise-free given then we're flipping between two uses of the same union (I know each effect type has its own m_effect) in an external API.  If house rules say don't bother, I'm fine removing it.  It's not my project.

**Comment by adamdmoss** [comment]
@karolherbst I'm really sorry I didn't see your tree, or I maybe wouldn't have bothered. :)  Though I think this implementation is complete and I've tested it with a few titles, I've only got an xbox 360 pad to test with.

**Comment by adamdmoss** [comment]
@karolherbst Bit surprised that SDL is letting you leave period as 0 in your branch and still doing something useful - I think that behaviour is undocumented.

**Comment by karolherbst** [comment]
I find the SDL_haptic docs pretty useless anyway :/ But I guess there are good enough default values if you keep values 0.

**Comment by karolherbst** [comment]
now I see, the peridic value is pushed through this macro internally: " #define  CLAMP(x)    (((x) > 32767) ? 32767 : x)"

**Comment by karolherbst** [comment]
ahh the period values is written into Linux' ff_effect.periodic.period value, so I think this is plattform specific and might be different on other plattforms. Maybe it is also driver specific.

**Comment by karolherbst** [comment]
What's weird though is, that you should notice a period of 1000ms :/ maybe I miss something here.

**Comment by adamdmoss** [comment]
@karolherbst I think you're right - the header examples do indicate that this is ms (don't you love unspecified-unit APIs?).  I'm guessing it's yet another thing that has implicit clamping and/or device-specific meaning.

**Comment by BhaaLseN** [comment]
Just another note on the memset, i merely assumed that it should-just-workTM. But then it should be added to the other effects aswell, if we really want to be on the safe side.

**Comment by adamdmoss** [comment]
@BhaaLseN Already done :)

**Comment by BhaaLseN** [comment]
@dolphin-emu-bot rebuild

**Comment by karolherbst** [comment]
I looked into the SDL haptic code and it seems like that at least on Linux all the stuff from the SDL structs is just copied over into kernel structs (even the effects are the same).

Also what do we want for the leftright effect? For me it still makes sense to split them or to find a combination, which feels like an original gc gamped.

**Comment by karolherbst** [comment]
@adamdmoss what about this? https://github.com/karolherbst/dolphin/commit/17ad68ff86c7508f4e8a82ae65f31bc2d59cee79

**Comment by adamdmoss** [comment]
I like it - I'd considered doing the same thing but just plain didn't.  I'll see if I can roughly cherry-pick that.

**Comment by karolherbst** [comment]
its based on your branch, so it should work

**Comment by karolherbst** [comment]
but honestly, I don't know if any compiler will complain about the "final" keyword, but a compiler with override support should also support final I guess.

**Comment by adamdmoss** [comment]
Thanks. :)  Refactored a little more on top.

**Comment by karolherbst** [comment]
mhh, I think it would be cleaner to squash all last three commits into one, because the commits just change stuff which the prior commits changed.

**Comment by karolherbst** [comment]
@adamdmoss  also what do you think about the leftright split? Maybe there are some who dislike a strong rumble and just want to stay with the little one. Or the other way around.

**Comment by BhaaLseN** [comment]
Since the effects do the same for cases where ```state``` isn't 0 (that is, setting ```m_effect.type = 0```), it might be worthwhile to think of cases where the the protected ```SetSDLHapticEffect``` wants to do something special - if there is none, just pull it out and only call it for non-zero, leaving the method to just setting appropriate values for a given control state.

**Comment by karolherbst** [comment]
also a good idea, makes the code in total much smaller than before :)

**Comment by adamdmoss** [comment]
I'll squash when/if it looks like we're close to acceptance.

I don't feel strongly about the left-right split, though IMHO Dolphin's haptic config UI is already going too far with its build-your-own-rumble interface - suspect most users just want any sort of rumble at all and perhaps a way to control magnitude (the slider doesn't seem hooked up in most/any backends as far as I can figure out).

I'll do that final refactor when I have a minute (unless someone wants to beat me to it :)) - I did notice it when I first looked at the implementation but figured I didn't want to mix a refactor with a feature addition, but we're well into that territory now. :)

**Comment by karolherbst** [comment]
yeah I understand what you say, but I think its good to have something like "small/big/both" as an option and nothing more, if every effect is the same in the end anyway.

Maybe that would be an idea to do: on gamepads with only one motor just enable it with a internal list of preferable effects and on gamepads with more motors, give the user the option to disable any of them, thus using only the leftright effect and nothing else.

**Comment by BhaaLseN** [comment]
Fixing the slider (should it be broken, I cannot really try this out at the moment) would be another PR, but refactoring is fine if it helps either code cleanliness, quality and/or future implementations. And on a somewhat personal preference, I like refactorings separate from actual changes (ie. two distinct commits) to make review a bit easier.

**Comment by karolherbst** [comment]
yeah, this was just an idea. I don't see the point of chaining up the effects together currently anyway. Is there any controller/effect combination where it does make a difference? This is something which should be discussed before anyway, because it might also change rumble handling for other plattforms.

**Comment by skidau** [comment]
@dolphin-emu-bot rebuild

**Comment by skidau** [comment]
@dolphin-emu-bot allowmerge

**Comment by karolherbst** [comment]
@adamdmoss I noticed, that sometimes rumble doesn't stop properly in Mario Party 5 (didn't test other games) and I had to wait until the next rumble effect comes. It might be an emulation problem.

**Comment by adamdmoss** [comment]
@karolherbst Interesting... might be an emulation problem, or an SDL/driver bug.  I guess we can use a non-infinite duration as a workaround, seems harmless to try.  Will add that after final refactor.

**Comment by karolherbst** [comment]
mhh maybe it is better to leave it there, because then the chances that it get fixed right is greater :D

**Comment by karolherbst** [comment]
but it should be noticed by others already, because the other effects have infinite durations. Strange enough the rumble effect continues after dolphin crashed, but this may be because it is actually kernel code.

**Comment by adamdmoss** [comment]
The other (existing) effects are much harder to notice if they never got stopped. :)  (Also probably useless for GC/WiiMote rumble emulation as they are, but I guess they're there for a reason and a user might get lucky.)

**Comment by adamdmoss** [comment]
I noticed too that rumble continues after an app finishes or crashes, if it's not explicitly turned off.  That's sadly not unique to Dolphin.

**Comment by adamdmoss** [comment]
(Dolphin does shut down the devices properly, if it didn't crash too catastrophically.)

**Comment by adamdmoss** [comment]
So barring any further concerns I'll squash and merge this in a few hours.  Thanks!

**Comment by karolherbst** [comment]
yeah, but I meant while running a game the rumble doesn't stop always. Its not that bad in someg ames, where the rumble is triggered a lot, but for others it may be a problem.

**Comment by karolherbst** [comment]
But I think I will retest this later and we could improve it later anyway. Its good to have rumble support at all.

**Comment by skidau** [comment]
@dolphin-emu-bot rebuild

**Comment by adamdmoss** [comment]
FYI, I got the sticky-rumble problem once more after several hours of play, it looks like an emulation bug since the finite duration is in place and SDL is still being update()d regularly (I presume - since pad input is obviously still working meanwhile).  Possibly an SDL bug but less likely since I've not observed it in other SDL titles.  Would be interested in whether this issue happens with Dolphin on other platforms.

**Comment by adamdmoss** [comment]
Believe this is ready to merge, though I don't have permission to do so.

**Comment by karolherbst** [comment]
@adamdmoss  https://github.com/dolphin-emu/dolphin/pull/1837#issuecomment-69481498 doesn't this work?

**Comment by adamdmoss** [comment]
Right:

Error: 403 Forbidden

Sorry, the requested URL 'http://localhost:8032/gh/merge/do/dolphin-emu/dolphin/1837/' caused an error:

Nobody allowed you to merge this PR

File: Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp [hunk 1]
```diff
@@ -18,6 +18,14 @@ namespace ciface
 namespace SDL
 {
 
+namespace
+{
+// 10ms = 100Hz which homebrew docs very roughly imply is within WiiMote normal
+// range, used for periodic haptic effects though often ignored by devices
+const u16 RUMBLE_PERIOD = 10;
+const u16 RUMBLE_LENGTH_MAX = 500; // ms: enough to span multiple frames at low FPS, but still finite
```
Comment (degasus) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp]: We mostly uses static insteat of anonymos namespaces, but I don't care much.

Comment (adamdmoss) [inline Source/Core/InputCommon/ControllerInterface/SDL/SDL.cpp]: That's fine, I'll change that - I don't care much myself, I was just copying what I saw in another random seemingly-well-written module. :)

**Comment by skidau** [comment]
@delroth does @adamdmoss need to be in the trusted developers group to merge?  See Error 403 above.

**Comment by skidau** [comment]
The rumble not stopping is a problem with all back-ends.  Dolphin should be sending a "stop rumble" command on shutdown.  Alternatively, the rumble period could be set to the current frame-time as measured in real time which varies depending on how fast the user's PC is.  This is not easily calculated.

**Comment by adamdmoss** [comment]
@skidau I don't think that's the rumble problem we're seeing, or at least the symptoms don't match - rumble stops okay on shutdown (except on bad crashes), but occasionally it sticks 'on' while playing games, stopping correctly again after a future rumble event.

**Comment by skidau** [comment]
@adamdmoss ah ok. I have not seen that myself (I use the xinput/xbox360 and the native gc adapter).

**Comment by skidau** [comment]
@dolphin-emu-bot rebuild

**Comment by jimbo1qaz** [comment]
Shouldn't it be possible to identify the frequency of the gamecube motor merely by listening to it?

**Comment by karolherbst** [comment]
I think it is much faster at max speed, and which motor do you mean, there are two of them.

**Comment by karolherbst** [comment]
@adamdmoss I will change some stuff now, because it doesn't feel that good, also I have still the rumble continues issue, which really breaks some mini games inside mario party 5. I try to find out why this happens and how we can make it less likely to occure if its not dolphins fault.

**Comment by adamdmoss** [comment]
Good.  I got a kernel panic (well, a trace in my dmesg anyway) that seemed to coincide with the effect staying on too long, and the trace had the usb driver at the top and the xbox pad driver underneath that.  Perhaps sending a haptic update every frame is flooding the device or the hub?  Instead of sending an update every frame we could try sending sending only when state actually flips.

**Comment by karolherbst** [comment]
my kernel didn't complain yet, even with two devices at the same time. But it seems likely to be a kernel bug, because SDL doesn't to much, maybe events get reordered for some reasons.

**Comment by karolherbst** [comment]
@adamdmoss : this patch fixes the rumbling issues: https://gist.github.com/karolherbst/8aaf997e5ecee260c262

**Comment by karolherbst** [comment]
but it is soo ugly, I think it is a SDL2 issue then

**Comment by karolherbst** [comment]
@adamdmoss  seems to be this SDL2 bug: https://bugzilla.libsdl.org/show_bug.cgi?id=1918

**Comment by adamdmoss** [comment]
Since June 2013 - that's depressing. :(

**Comment by karolherbst** [comment]
@adamdmoss  https://bugzilla.kernel.org/show_bug.cgi?id=91991 I've opened a kernel bug for it today, because with pure Linux API you can trigger it, too

**Comment by adamdmoss** [comment]
Good work!  How reliable is the repro?

**Comment by karolherbst** [comment]
very. It happens like 50% of the times I try the kernel code and 100% with the SDL code. Allthough both codes cover different scenarios. You can try out yourself if you want. Just run them more often.

The problem with the SDL code is, that an update followed by a stop may cause this issue, but if enough time is spent between both calls it does not happen. I assume that something is wrong with the force feedback API, because you use write() for update and ioctrl() for stop, and the write call may delay the actual update or trigger a race condition inside kernel code, which push the driver into invalid state.

**Pull request merged**